core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java

```diff
@@ -0,0 +1,810 @@
+      try {
+        mRootPath.traverse();
+      } catch (InvalidPathException e) {
+        throw new RuntimeException(e);
+      }
+    } catch (AccessControlException | BlockInfoException | FileAlreadyCompletedException
+        | FileDoesNotExistException | InvalidFileSizeException | InvalidPathException
+        | IOException e) {
+      LOG.warn("FAILED TO SYNC METADATA: {}", e.getMessage(), e);
+    } finally {
+      // regardless of the outcome, remove the UfsStatus for this path from the cache
+      mStatusCache.remove(mRootPath.getUri());
+      // downgrade so that if operations are parallelized, the lock on the root doesn't restrict
+      // concurrent operations
+      mRootPath.downgradeToPattern(LockPattern.READ);
+    }
+
+    // Process any children after the root.
+    while (!mSyncMetadataQ.isEmpty() || !mSyncPathJobs.isEmpty()) {
+      if (Thread.currentThread().isInterrupted()) {
+        LOG.warn("Metadata syncing was interrupted before completion");
+        break;
+      }
+      // There are still paths to process
+      // First, remove any futures which have completed. Add to the sync path count if they sync'd
+      // successfully
+      while (true) {
+        Future<Boolean> job = mSyncPathJobs.peek();
+        if (job == null || !job.isDone()) {
+          break;
+        }
+        // remove the job because we know it is done.
+        if (mSyncPathJobs.poll() != job) {
+          throw new IllegalStateException("Last node to be de-queued was not equal to the expected"
+              + "head of queue");
+        }
+        try {
+          // we synced the path successfully
+          if (job.get()) {
+            syncPathCount++;
+          }
+        } catch (InterruptedException | ExecutionException e) {
+          if (e instanceof  InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          LOG.warn("metadata sync job was interrupted while waiting for completion");
+        }
+      }
+
+      // When using descendant type of ONE, we need to stop prematurely.
+      if (stopNum != -1 && syncPathCount > stopNum) {
+        break;
+      }
+
+      // We can submit up to ( max_concurrency - <jobs queue size>) jobs back into the queue
+      int submissions = mConcurrencyLevel - mSyncPathJobs.size();
+      for (int i = 0; i < submissions; i++) {
+        AlluxioURI path = mSyncMetadataQ.poll();
+        if (path == null) {
+          // no paths left to sync
+          break;
+        }
+        Future<Boolean> job = mMetadataSyncService.submit(() -> processSyncPath(path));
+        mSyncPathJobs.offer(job);
+      }
+      // After submitting all jobs wait for the job at the head of the queue to finish.
+      Future<Boolean> oldestJob = mSyncPathJobs.peek();
+      if (oldestJob == null) { // There might not be any jobs, restart the loop.
+        continue;
+      }
+      try {
+        oldestJob.get(); // block until the oldest job finished.
```

Comment:
Does it matter if the oldest job finishes first or last? How would that affect behavior?

Reply:
I just take the oldest job submitted, because as long as the threadpool uses a fair queueing policy for running jobs, it has the highest probability of being completed before other jobs. If the oldest job runs slower, it can theoretically prevent other tasks from being submitted.

In a follow-up I'd actually like to design a ThreadPoolExecutor that doesn't need this control logic and can simply just iterate over the queue of pending paths and submit them. My idea is that the executor pool can be shared in a limited manner, so that you can lease out a portion of the threads. ex:

```java
ExecutorService mainExecutor = Executors.newCachedThreadpool(64);

sharedExecutor = mainExecutor.lease(4); // executor service that uses the pool backing mainExecutor, that can have up to 4 tasks running concurrently.
sharedExecutor = mainExecutor.lease(32); // executor service that uses the pool backing mainExecutor, that can have up to 32 tasks running concurrently.
```

We can use a semaphore to block submitting tasks to the executor when they are already runing. Completely negates all the control logic of this thread. I haven't written the executor yet. This is all ideas yet. I don't plan on implementing it for this PR